# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/ipc/CanvasManagerParent.cpp
# Commit: dab6b48a2a59
# Full Hash: dab6b48a2a592a441229e88c88b1aa2570f2171d
# Author: Andrew Osmond <aosmond@mozilla.com>
# Date: 2023-12-12 04:37:49
# Regressor Bug: 1869346
# File Overlap Count: 5
# Description:
#   Bug 1869346 - Rework how canvas accesses shared surfaces in the compositor process. r=gfx-reviewers,lsalzman
#   
#   This patch adds plumbing to allow sharing ExternalImageId via
#   SurfaceDescriptor. This will be used in a future patch by WebGL and
#   WebGPU to avoid extra copies. It also refactors how CanvasTranslator
# ==============================================================================

diff -r af13afc7d44f -r dab6b48a2a59 gfx/ipc/CanvasManagerParent.cpp
--- a/gfx/ipc/CanvasManagerParent.cpp	Mon Dec 11 20:19:47 2023 +0000
+++ b/gfx/ipc/CanvasManagerParent.cpp	Mon Dec 11 20:44:44 2023 +0000
@@ -14,6 +14,7 @@
 #include "mozilla/layers/CanvasTranslator.h"
 #include "mozilla/layers/CompositorThread.h"
 #include "mozilla/layers/ISurfaceAllocator.h"
+#include "mozilla/layers/SharedSurfacesParent.h"
 #include "mozilla/webgpu/WebGPUParent.h"
 #include "nsIThread.h"
 #include "nsThreadUtils.h"
@@ -29,10 +30,12 @@
 
 /* static */ void CanvasManagerParent::Init(
     Endpoint<PCanvasManagerParent>&& aEndpoint,
+    layers::SharedSurfacesHolder* aSharedSurfacesHolder,
     const dom::ContentParentId& aContentId) {
   MOZ_ASSERT(layers::CompositorThreadHolder::IsInCompositorThread());
 
-  auto manager = MakeRefPtr<CanvasManagerParent>(aContentId);
+  auto manager =
+      MakeRefPtr<CanvasManagerParent>(aSharedSurfacesHolder, aContentId);
 
   nsCOMPtr<nsIThread> owningThread =
       gfx::CanvasRenderThread::GetCanvasRenderThread();
@@ -220,8 +223,10 @@
   return desc;
 }
 
-CanvasManagerParent::CanvasManagerParent(const dom::ContentParentId& aContentId)
-    : mContentId(aContentId) {}
+CanvasManagerParent::CanvasManagerParent(
+    layers::SharedSurfacesHolder* aSharedSurfacesHolder,
+    const dom::ContentParentId& aContentId)
+    : mSharedSurfacesHolder(aSharedSurfacesHolder), mContentId(aContentId) {}
 
 CanvasManagerParent::~CanvasManagerParent() = default;
 
@@ -273,7 +278,8 @@
 already_AddRefed<layers::PCanvasParent>
 CanvasManagerParent::AllocPCanvasParent() {
   MOZ_RELEASE_ASSERT(mId != 0);
-  return MakeAndAddRef<layers::CanvasTranslator>(mContentId, mId);
+  return MakeAndAddRef<layers::CanvasTranslator>(mSharedSurfacesHolder,
+                                                 mContentId, mId);
 }
 
 mozilla::ipc::IPCResult CanvasManagerParent::RecvGetSnapshot(